### PR TITLE
feat: vLLM inference backend for open-source model deployment

### LINKED ISSUE #812: feat: vLLM inference backend for open-source model deployment
## Summary

Add vLLM as a high-throughput self-hosted inference backend, filling the gap between Ollama (easy, slow) and cloud APIs (fast, costly). PagedAttention gives 20–24× higher throughput vs HuggingFace Transformers - critical for batch geospatial raster inference.

## Motivation

Current LLM backends (`create_ollama_model`, `create_openai_model`, etc.) don't support high-throughput local serving of larger open-source VLMs. Researchers running GPU clusters or cloud VMs need a way to serve models like Qwen2-VL or LLaVA at scale without API costs.

### Layer 1 - Agent text LLM (minimal)

Add `create_vllm_model()` to `geoai/agents/geo_agents.py`. Since vLLM exposes an OpenAI-compatible REST API, this wraps the existing strands `OpenAIModel` with a custom `base_url`:

```python
def create_vllm_model(
    base_url: str = "http://localhost:8000/v1",
    model_id: str = "meta-llama/Llama-3.1-8B-Instruct",
    api_key: str = "EMPTY",
): ...
```

### Layer 2 - VLM geospatial inference class

New `geoai/vllm_geo.py` mirroring `moondream.py` structure:

```python
class VLLMGeo:
    def caption(source, prompt, ...)        -> str
    def query(question, source, ...)        -> str
    def detect(source, object_type, ...)    -> GeoDataFrame
    def caption_sliding_window(source, ...)
    def query_sliding_window(source, ...)
```

Two modes:
- **Server mode** (default): calls `vllm serve <model>` OpenAI endpoint with base64-encoded image tiles
- **In-process mode** (`offline=True`): loads via `vllm.LLM` directly

Convenience functions: `vllm_caption()`, `vllm_query()`

## Suggested Supported Models

| Model | HuggingFace ID | Use case |
|-------|---------------|----------|
| Qwen2-VL-7B | `Qwen/Qwen2-VL-7B-Instruct` | Caption + VQA (best quality) |
| LLaVA-1.6 | `llava-hf/llava-v1.6-mistral-7b-hf` | General VQA |
| InternVL2-8B | `OpenGVLab/InternVL2-8B` | Scene understanding |
| Llama-3.1-8B | `meta-llama/Llama-3.1-8B-Instruct` | Text agent LLM |

## References

- vLLM docs: https://docs.vllm.ai
- vLLM vision model support: https://docs.vllm.ai/en/latest/models/vision_language.html
- PagedAttention paper: https://arxiv.org/abs/2309.06180
- Qwen2-VL: https://huggingface.co/Qwen/Qwen2-VL-7B-Instruct
- Existing pattern to mirror: `geoai/moondream.py`, `geoai/agents/geo_agents.py`

## Implementation Notes

- Reuse sliding window logic from `moondream.py`
- Reuse `utils/device.py:get_device()` for in-process mode
- Mirror `check_terratorch_available()` pattern for `check_vllm_available()`
- Default to server mode; in-process as opt-in via `offline=True`

## Good First Issue

Layer 1 (`create_vllm_model()`) is ~20 lines and a great entry point. Layer 2 can follow independently.

### PR DESCRIPTION
**Pull request by jayakumarpujar**
## Summary

Implements #812 — adds vLLM as a high-throughput self-hosted inference backend, filling the gap between Ollama (easy, slow) and cloud APIs (fast, costly).

### Layer 1 — Agent text LLM
- `create_vllm_model()` in `geoai/agents/geo_agents.py`: wraps the strands `OpenAIModel` with a custom `base_url` pointing at a vLLM OpenAI-compatible server (mirrors the `create_minimax_model` pattern). Defaults: `http://localhost:8000/v1`, `meta-llama/Llama-3.1-8B-Instruct`, `api_key="EMPTY"`.

### Layer 2 — VLM geospatial inference
- New `geoai/vllm_geo.py` mirroring `moondream.py`:
  - `VLLMGeo.caption()` / `query()` / `detect()` — detect uses JSON-prompted bounding boxes (normalized 0–1), parsed robustly and georeferenced to a GeoDataFrame with vector export (GeoJSON/SHP/GPKG).
  - `caption_sliding_window()` / `query_sliding_window()` / `detect_sliding_window()` — tiled processing for large rasters with NMS merging and concatenate/summarize combine strategies.
  - **Server mode** (default): calls the `vllm serve` OpenAI endpoint with base64 image tiles — only needs `requests`, no vLLM install required locally.
  - **In-process mode** (`offline=True`): loads via `vllm.LLM` directly.
- `check_vllm_available()` mirroring `check_terratorch_available()`.
- Convenience functions: `vllm_caption()`, `vllm_query()`, `vllm_detect()`.
- Lazy top-level exports in `geoai/__init__.py`, `vllm` optional-dependency extra, `docs/vllm_geo.md` API page.

## Test plan
- [x] 44 unit tests in `tests/test_vllm_geo.py` (mocked server responses, no GPU/network needed): image loading (PIL/numpy/GeoTIFF), base64 encoding, caption/query/detect, JSON parsing edge cases (surrounding text, malformed, degenerate/out-of-range boxes), georeferencing, vector export, sliding windows + NMS, convenience functions, `create_vllm_model()` config (skipped when strands not installed)
- [x] `pytest tests/test_vllm_geo.py` — 41 passed, 3 skipped locally
- [x] black formatting applied
- [ ] Manual smoke test against a live `vllm serve Qwen/Qwen2-VL-7B-Instruct` server (requires GPU)

Closes #812

## Summary by CodeRabbit

* **New Features**
  * Added vLLM-based geospatial vision-language support for captioning, visual question answering, and object detection, with server or in-process execution and convenience wrapper functions.
  * Supports GeoTIFF, images, and arrays; detection parsing, optional georeferencing, GeoDataFrame/GeoJSON output, and sliding-window tiling with tile merging.
  * Exposed new lazy-import APIs and an agent model factory for OpenAI-compatible vLLM endpoints.
* **Documentation**
  * Added documentation page for the new `vllm_geo` module.
* **Optional Dependencies**
  * Introduced a `vllm` installation extra.
* **Tests**
  * Added end-to-end unit tests with mocked server calls and GeoTIFF fixtures.

### CODE REVIEW

File: geoai/vllm_geo.py [hunk 1]
```diff
@@ -0,0 +1,1105 @@
...
+            return self.detect(image, object_type, output_path=output_path, **kwargs)
+
+        windows = self._create_sliding_windows(width, height, window_size, overlap)
+        all_detections = []
+
+        iterator = (
+            tqdm(windows, desc=f"Detecting {object_type}") if show_progress else windows
+        )
+
+        prompt = DETECT_PROMPT_TEMPLATE.format(object_type=object_type)
```
[the remark this answers is not in the document]
Comment (giswqs) [reply geoai/vllm_geo.py]: Done in 30012a98 — confirmed: `load_image` returns `metadata=None` for PIL input (geoai/vllm_geo.py:226-229), so the short-circuit lost CRS/transform and never wrote `output_path`. The small-image branch now inlines `_chat`/`_parse_detections` and reuses the metadata loaded at the top of `detect_sliding_window`, so it georeferences and saves exactly like the tiled path.

File: geoai/agents/geo_agents.py [hunk 2]
```diff
@@ -287,6 +287,46 @@ def create_minimax_model(
     return OpenAIModel(client_args=client_args, model_id=model_id, **kwargs)
 
 
+def create_vllm_model(
+    base_url: str = "http://localhost:8000/v1",
+    model_id: str = "meta-llama/Llama-3.1-8B-Instruct",
+    api_key: str = "EMPTY",
+    client_args: dict = None,
+    **kwargs: Any,
```
[the remark this answers is not in the document]
Comment (giswqs) [reply geoai/agents/geo_agents.py]: Partially done in 30012a98.

**RUF013 (implicit Optional)** — fixed: `client_args` is now `Optional[dict]`, which matches the idiom already used elsewhere in this file (e.g. `map_instance: Optional[leafmap.Map]`).

**ANN401 (`**kwargs: object`)** — declining. `kwargs` is forwarded verbatim to `OpenAIModel(**kwargs)`, whose own signature is `Any`-typed; narrowing to `object` would be less accurate for callers, not more. `**kwargs: Any` is also the uniform convention across all six `create_*_model` factories in this module. Note this repo has no Ruff config — the CI gate is black + codespell + bandit — so neither rule is enforced here.
